### PR TITLE
add type annotations to torch.nn.modules.fold

### LINKED ISSUE #49478: Enable torch.nn.modules.fold typechecks during CI
`TestTypeHints.test_run_mypy` should pass with `[mypy-torch.nn.modules.fold]` deleted from mypy.ini

cc @ezyang @malfet @rgommers @xuzhao9 @gramster

### PR DESCRIPTION
**Pull request by guilhermeleobas**
closes gh-49478

Fixes #49478

### CODE REVIEW

**Review by walterddr — APPROVED** [review]

**Review by rgommers — APPROVED** [review]
LGTM, and one CI failure is unrelated. Thanks @guilhermeleobas

**Review by malfet — APPROVED** [review]

**Comment by guilhermeleobas** [comment]
Do not merge this PR until one checks if the annotations introduce any regression. See:
https://github.com/pytorch/pytorch/pull/49564#issuecomment-752916072

**Comment by guilhermeleobas** [comment]
This module is already covered by tests:
```
test/test_jit.py::TestJitGeneratedModule::test_nn_Fold
test/test_jit.py::TestJitGeneratedModule::test_nn_Fold_int_input
test/test_jit.py::TestJitGeneratedModule::test_nn_Unfold
test/test_jit.py::TestJitGeneratedModule::test_nn_Unfold_int_input
test/test_jit.py::TestJitGeneratedFunctional::test_nn_fold
test/test_jit.py::TestJitGeneratedFunctional::test_nn_unfold
```

**Pull request closed without merging**
